lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -660,6 +642,104 @@ bool mlir::triton::preProcessLoopAndGetSchedule(
   return true;
 }
 
+/// Find the minimum number of insert_slice ops between the extract
+/// and the insert. This is equivalent to counting async_commit_group ops
+/// (unless we will start emitting multiple insert_slice ops per commit_group).
+static int minWaitNumberForExtract(ttg::ExtractSliceOp extractOp) {
+  auto countInsertsBetween = [](Operation *op1, Operation *op2) {
+    int count = 0;
+    for (auto op = op1; op != op2; op = op->getNextNode()) {
+      if (isa<ttg::InsertSliceAsyncOp>(op))
+        count++;
+      // Intentionally skip block ops' children. This will give us
+      // convervatively low number of insert ops.
+    }
+    return count;
+  };
+
+  int minWaitNumber = INT_MAX;
+
+  // DFS the def chain of the extract op to find the insert op. On each path
+  // we calculate the number of insert ops. Then we select the minimum number
+  // of insert ops among all the paths.
+  // If the wait is not needed (when insert is a TMA insert), we return -1.
+  std::function<int(Value, Operation *, int)> minOverHistories =
+      [&](Value val, Operation *sinkOp, int thisHistorySum) -> int {
+    if (auto defOp = val.getDefiningOp()) {
+      if (isa<ttg::InsertSliceAsyncOp>(defOp)) {
+        thisHistorySum += countInsertsBetween(defOp->getNextNode(), sinkOp);
+        minWaitNumber = std::min(minWaitNumber, thisHistorySum);
+        return minWaitNumber;
+      }
+      if (isa<ttng::InsertSliceTMAOp>(defOp)) {
+        // We don't need to wait for TMA inserts.
+        return -1;
+      }
```

Comment:
so if there is a mix of `InsertSliceTMAOp` and others we will end up with -1, this doesn't sound right, can we just skip this case or return `INT_MAX` saying basically this shouldn't affect the count?

Reply:
Is it possible that single extract will be sourced by TMA and non-TMA insert?

Reply:
it's not my bad